### ISSUE TITLE
regression: "unexpected missing blob" in "Checks + restart Redpanda & Debezium" Nightly

### ISSUE THREAD

**Issue by lluki** [MaterializeInc/materialize#17338]
### Buildkite link

https://buildkite.com/materialize/tests/builds/48896

### Relevant log output

```source
> SELECT f3 FROM create_index_table;
rows match; continuing at ts 1674643397.6186798
Running validate() from <materialize.checks.create_table.CreateTable object at 0x7f2cc38b8c10>
$ docker compose exec -T testdrive testdrive --kafka-addr=kafka:9092 --schema-registry-url=http://schema-registry:8081 --materialize-url=postgres://materialize@materialized:6875 --materialize-internal-url=postgres://materialize@materialized:6877 --aws-endpoint=http://localstack:4566 --no-reset --default-timeout=300s --seed=1
> SELECT * FROM create_table1;
rows match; continuing at ts 1674643399.5517387
> SELECT * FROM create_table2;
rows match; continuing at ts 1674643404.1382248
> SELECT * FROM create_table3;
rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 49s 74s 68s
8:1: error: preparing query failed: connection closed
     |
   7 |
   8 | > SELECT * FROM create_table3;
     | ^
!!! Error Report
```

Happens because a compute replica crashes:

```
platform-checks-materialized-1  | cluster-u11-replica-13: 2023-01-25T10:43:31.416858Z  INFO mz_storage::source::source_reader_pipeline: reclock(u82) 1/4: produced new compaction since for the reclock operator compaction_since=Antichain { elements: [1674643411000] } resumption_frontier=Antichain { elements: [1674643411001] }
platform-checks-materialized-1  | cluster-u11-replica-13: 2023-01-25T10:43:31.416545Z  INFO mz_storage::source::source_reader_pipeline: reclock(u82) 3/4: calculated offset commit frontier for resumption frontier update resumption_frontier=Antichain { elements: [1674643411001] } offset_upper=OffsetAntichain { inner: {} }
platform-checks-materialized-1  | cluster-u11-replica-13: 2023-01-25T10:43:31.417068Z  INFO mz_storage::source::source_reader_pipeline: reclock(u82) 3/4: produced new compaction since for the reclock operator compaction_since=Antichain { elements: [1674643411000] } resumption_frontier=Antichain { elements: [1674643411001] }
platform-checks-materialized-1  | cluster-u11-replica-13: 2023-01-25T10:43:31.416515Z  INFO mz_storage::source::source_reader_pipeline: reclock(u82) 0/4: calculated offset commit frontier for resumption frontier update resumption_frontier=Antichain { elements: [1674643411001] } offset_upper=OffsetAntichain { inner: {} }
platform-checks-materialized-1  | cluster-u11-replica-13: 2023-01-25T10:43:31.417333Z  INFO mz_storage::source::source_reader_pipeline: reclock(u82) 0/4: produced new compaction since for the reclock operator compaction_since=Antichain { elements: [1674643411000] } resumption_frontier=Antichain { elements: [1674643411001] }
platform-checks-materialized-1  |    0: rust_begin_unwind
platform-checks-materialized-1  |              at ./rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:575:5
platform-checks-materialized-1  |    1: core::panicking::panic_fmt
platform-checks-materialized-1  |              at ./rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:65:14
platform-checks-materialized-1  |    2: mz_persist_client::fetch::fetch_leased_part::<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>::{closure#0}::{closure#3}
platform-checks-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0dfa9a7c485d41ae0-1/materialize/tests/src/persist-client/src/fetch.rs:205:9
platform-checks-materialized-1  |    3: <core::result::Result<mz_persist_client::fetch::EncodedPart<mz_repr::timestamp::Timestamp>, anyhow::Error>>::unwrap_or_else::<mz_persist_client::fetch::fetch_leased_part<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>::{closure#0}::{closure#3}>
platform-checks-materialized-1  |              at ./rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/result.rs:1504:23
platform-checks-materialized-1  |    4: mz_persist_client::fetch::fetch_leased_part::<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>::{closure#0}
platform-checks-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0dfa9a7c485d41ae0-1/materialize/tests/src/persist-client/src/fetch.rs:187:24
platform-checks-materialized-1  |    5: <core::future::from_generator::GenFuture<mz_persist_client::fetch::fetch_leased_part<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>::{closure#0}> as core::future::future::Future>::poll
platform-checks-materialized-1  |              at ./rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/future/mod.rs:91:19
platform-checks-materialized-1  |    6: <mz_persist_client::fetch::BatchFetcher<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>>::fetch_leased_part::{closure#0}
platform-checks-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0dfa9a7c485d41ae0-1/materialize/tests/src/persist-client/src/fetch.rs:109:9
platform-checks-materialized-1  |    7: <core::future::from_generator::GenFuture<<mz_persist_client::fetch::BatchFetcher<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>>::fetch_leased_part::{closure#0}> as core::future::future::Future>::poll
platform-checks-materialized-1  |              at ./rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/future/mod.rs:91:19
platform-checks-materialized-1  |    8: mz_persist_client::operators::shard_source::shard_source_fetch::<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}::{closure#0}
platform-checks-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0dfa9a7c485d41ae0-1/materialize/tests/src/persist-client/src/operators/shard_source.rs:486:25
platform-checks-materialized-1  |    9: <core::future::from_generator::GenFuture<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}::{closure#0}> as core::future::future::Future>::poll
platform-checks-materialized-1  |              at ./rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/future/mod.rs:91:19
platform-checks-materialized-1  |   10: <core::pin::Pin<alloc::boxed::Box<core::future::from_generator::GenFuture<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}::{closure#0}>>> as core::future::future::Future>::poll
platform-checks-materialized-1  |              at ./rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/future/future.rs:124:9
platform-checks-materialized-1  |   11: <mz_timely_util::builder_async::OperatorBuilder<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>>::build::<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}, core::future::from_generator::GenFuture<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}::{closure#0}>>::{closure#0}::{closure#0}
platform-checks-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0dfa9a7c485d41ae0-1/materialize/tests/src/timely-util/src/builder_async.rs:326:28
platform-checks-materialized-1  |   12: <timely::dataflow::operators::generic::builder_rc::OperatorBuilder<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>>::build_reschedule::<<mz_timely_util::builder_async::OperatorBuilder<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>>::build<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}, core::future::from_generator::GenFuture<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}::{closure#0}>>::{closure#0}, <mz_timely_util::builder_async::OperatorBuilder<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>>::build<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}, core::future::from_generator::GenFuture<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}::{closure#0}>>::{closure#0}::{closure#0}>::{closure#0}
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/dataflow/operators/generic/builder_rc.rs:162:26
platform-checks-materialized-1  |   13: <timely::dataflow::operators::generic::builder_raw::OperatorCore<mz_repr::timestamp::Timestamp, <timely::dataflow::operators::generic::builder_rc::OperatorBuilder<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>>::build_reschedule<<mz_timely_util::builder_async::OperatorBuilder<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>>::build<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}, core::future::from_generator::GenFuture<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}::{closure#0}>>::{closure#0}, <mz_timely_util::builder_async::OperatorBuilder<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>>::build<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}, core::future::from_generator::GenFuture<mz_persist_client::operators::shard_source::shard_source_fetch<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64, timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>>::{closure#1}::{closure#0}>>::{closure#0}::{closure#0}>::{closure#0}> as timely::scheduling::Schedule>::schedule
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/dataflow/operators/generic/builder_raw.rs:204:9
platform-checks-materialized-1  |   14: <timely::progress::subgraph::PerOperatorState<mz_repr::timestamp::Timestamp>>::schedule
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/progress/subgraph.rs:684:30
platform-checks-materialized-1  |   15: <timely::progress::subgraph::Subgraph<mz_repr::timestamp::Timestamp, mz_repr::timestamp::Timestamp>>::activate_child
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/progress/subgraph.rs:344:26
platform-checks-materialized-1  |   16: <timely::progress::subgraph::Subgraph<mz_repr::timestamp::Timestamp, mz_repr::timestamp::Timestamp> as timely::scheduling::Schedule>::schedule
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/progress/subgraph.rs:310:17
platform-checks-materialized-1  |   17: <timely::progress::subgraph::PerOperatorState<mz_repr::timestamp::Timestamp>>::schedule
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/progress/subgraph.rs:684:30
platform-checks-materialized-1  |   18: <timely::progress::subgraph::Subgraph<mz_repr::timestamp::Timestamp, mz_repr::timestamp::Timestamp>>::activate_child
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/progress/subgraph.rs:344:26
platform-checks-materialized-1  |   19: <timely::progress::subgraph::Subgraph<mz_repr::timestamp::Timestamp, mz_repr::timestamp::Timestamp> as timely::scheduling::Schedule>::schedule
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/progress/subgraph.rs:310:17
platform-checks-materialized-1  |   20: <timely::progress::subgraph::PerOperatorState<mz_repr::timestamp::Timestamp>>::schedule
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/progress/subgraph.rs:684:30
platform-checks-materialized-1  |   21: <timely::progress::subgraph::Subgraph<(), mz_repr::timestamp::Timestamp>>::activate_child
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/progress/subgraph.rs:344:26
platform-checks-materialized-1  |   22: <timely::progress::subgraph::Subgraph<(), mz_repr::timestamp::Timestamp> as timely::scheduling::Schedule>::schedule
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/progress/subgraph.rs:310:17
platform-checks-materialized-1  |   23: <timely::worker::Wrapper>::step::{closure#0}
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/worker.rs:754:57
platform-checks-materialized-1  |   24: <core::option::Option<&mut alloc::boxed::Box<dyn timely::scheduling::Schedule>>>::map::<bool, <timely::worker::Wrapper>::step::{closure#0}>
platform-checks-materialized-1  |              at ./rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/option.rs:925:29
platform-checks-materialized-1  |   25: <timely::worker::Wrapper>::step
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/worker.rs:754:26
platform-checks-materialized-1  |   26: <timely::worker::Worker<timely_communication::allocator::generic::Generic>>::step_or_park
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/worker.rs:396:38
platform-checks-materialized-1  |   27: <mz_compute::server::Worker<timely_communication::allocator::generic::Generic>>::run_client
platform-checks-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0dfa9a7c485d41ae0-1/materialize/tests/src/compute/src/server.rs:553:13
platform-checks-materialized-1  |   28: <mz_compute::server::Worker<timely_communication::allocator::generic::Generic>>::setup_channel_and_run_client
platform-checks-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0dfa9a7c485d41ae0-1/materialize/tests/src/compute/src/server.rs:527:9
platform-checks-materialized-1  |   29: <mz_compute::server::Worker<timely_communication::allocator::generic::Generic>>::run
platform-checks-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0dfa9a7c485d41ae0-1/materialize/tests/src/compute/src/server.rs:387:21
platform-checks-materialized-1  |   30: <mz_compute::server::ClusterClient<mz_service::client::Partitioned<mz_service::local::LocalClient<mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse, timely::scheduling::activate::SyncActivator>, mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse>>>::build_timely::{closure#0}::{closure#1}
platform-checks-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0dfa9a7c485d41ae0-1/materialize/tests/src/compute/src/server.rs:184:13
platform-checks-materialized-1  |   31: timely::execute::execute_from::<timely_communication::allocator::generic::GenericBuilder, (), <mz_compute::server::ClusterClient<mz_service::client::Partitioned<mz_service::local::LocalClient<mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse, timely::scheduling::activate::SyncActivator>, mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse>>>::build_timely::{closure#0}::{closure#1}>::{closure#0}
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/timely/src/execute.rs:388:22
platform-checks-materialized-1  |   32: timely_communication::initialize::initialize_from::<timely_communication::allocator::generic::GenericBuilder, (), timely::execute::execute_from<timely_communication::allocator::generic::GenericBuilder, (), <mz_compute::server::ClusterClient<mz_service::client::Partitioned<mz_service::local::LocalClient<mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse, timely::scheduling::activate::SyncActivator>, mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse>>>::build_timely::{closure#0}::{closure#1}>::{closure#0}>::{closure#0}
platform-checks-materialized-1  |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/faf5eb6/communication/src/initialize.rs:316:33
platform-checks-materialized-1  | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
platform-checks-materialized-1  | cluster-u1-replica-1: 2023-01-25T10:43:31.883537Z  WARN mz_timely_util::panic: halting process: timely communication error: reading data: Connection reset by peer (os error 104)
platform-checks-materialized-1  | cluster-u1-replica-1: 2023-01-25T10:43:31.885870Z  WARN mz_timely_util::panic: halting process: timely communication error: reading data: Connection reset by peer (os error 104)
platform-checks-materialized-1  | cluster-u1-replica-1: 2023-01-25T10:43:31.885879Z  WARN mz_timely_util::panic: halting process: timely communication error: reading data: Connection reset by peer (os error 104)
platform-checks-materialized-1  | thread 'tokio:work-2' panicked at 'cluster-u1-replica-1-2 crashed; aborting because propagate_crashes is enabled', src/orchestrator-process/src/lib.rs:558:29
platform-checks-materialized-1  | stack backtrace:
platform-checks-materialized-1  | environmentd: 2023-01-25T10:43:31.895737Z  WARN mz_compute_client::controller::replica: replica task for 1 failed: status: Unknown, message: "error reading a body from connection: broken pipe", details: [], metadata: MetadataMap { headers: {} }
```

### Additional thoughts

There are plenty of concerning log messages:

```
platform-checks-materialized-1  | cluster-u14-replica-16: 2023-01-25T10:41:55.185069Z  WARN mz_storage::sink::kafka: Kafka producer delivery error KafkaError (Message production error: PurgeQueue (Local: Purged in queue)) for Message { ptr: 0x7fb235293600 }
```

I think the root cause is that something goes wrong with Kafka, ie. this happens before the above crash:

```
platform-checks-materialized-1  | cluster-u3-replica-5: 2023-01-25T10:41:50.553937Z ERROR librdkafka: Global error: AllBrokersDown (Local: All broker connections are down): 2/2 brokers are down
```

ci-regexp: unexpected missing blob

**Comment by aljoscha** [comment]
@pH14 the panic seems to be your beloved  batch fetcher problem: https://github.com/MaterializeInc/materialize/issues/17338#issuecomment-1404793703 :see_no_evil:

**Comment by pH14** [comment]
This is most likely a known bug where unclean shutdown can cause a `shard_source` to panic, and we see a small trickle of them in Sentry. When the dataflow is dropped, the worker that owns the `Subscribe` may get dropped first, meaning its seqno hold is released, and other workers may still be trying to fetch the parts it emitted before they are dropped. If GC deletes the blobs before the fetches happen, we hit this panic.

We've tossed about a few options for resolving it, but none super satisfying. The quickest solution would be dropped the `Subscribe` without expiring the reader in state, so downstream workers have 15min to finish fetching their batches, but I haven't liked that option enough to type it up yet.

**Comment by danhhz** [comment]
> The quickest solution would be dropped the Subscribe without expiring the reader in state, so downstream workers have 15min to finish fetching their batches, but I haven't liked that option enough to type it up yet.

I think it's time for us to type this up. At the very least to hold us over until/if we think of something better.

**Comment by pH14** [comment]
My concern is that it introduces more complications that are subtle to debug when they occur. It plays poorly with https://github.com/MaterializeInc/materialize/pull/17147 because we're intentionally leaving workers to fetch batches after a dataflow is dropped, and leaving straggler read handles around can affect our ability to compact because we won't advance since for 15m after each one is dropped. Our current compaction setup doesn't recover well when we belatedly advance since because we might physically compact data without logical compaction, and it could take a long/indefinite time to revisit those batches. Relying on the 15m window each time seems like it could be pathologically bad for shards that receive any amount of ad-hoc queries because `since` will always run 15m behind at best. Not to say we shouldn't do it, but IMO it's not without downsides (beyond just feeling a bit icky) that are worth considering

**Comment by danhhz** [comment]
Yeah, okay that's a fantastic point! Great framing. One thing that's interesting about it is the ReadHandle has two separate capabilities: one on since and one on seqno. The compaction drawbacks you point out only apply to leaking the since capability, but the only thing we want to leak here is the seqno capability. That only affects how long gc is held back and between (1) occasional panics with a scary message that results in process restarts and (2) leaving data in blob and consensus up to 15m longer than strictly necessary, it seems like it's an easy call for (2) as the interim behavior.

As for actually effecting that, one idea (not sure if I love this) is to downgrade_since to empty (not via the `maybe-` variant) for the ReadHandle before intentionally leaking the registration. One immediate concern with that is that expire_reader is guaranteed to never make the since advance (a holdover from the days before critical handles where we were patching things to buy time) but downgrade_since very much could. AFAICT, that shouldn't matter anymore now that we have critical handles, but might be worth thinking through.

**Comment by philip-stoev** [comment]
Happened today on the "Product Limits" Nightly test on a branch.

**Comment by umanwizard** [comment]
I reproduced this in Staging today.

**Comment by guswynn** [comment]
This job has been very flaky recently, I have seen it also fail with "agent lost" and random "connection close" errors, which makes me think that something like an oom or some other weird problem is happening

**Comment by philip-stoev** [comment]
No, The product limits test is failing explicitly with:

```
limits-materialized-1     | thread 'timely:work-0' panicked at 'batch fetcher could not fetch batch part: unexpected missing blob: w5443f4e3-05fa-4ee2-b89d-fc497988f4fb/p3223fc55-9215-4801-803f-110f6da42d99 for shard: s0025f4bb-8a8f-4b00-a1f4-6fc7fa31b230', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0616339172b8538fc-1/materialize/tests/src/persist-client/src/fetch.rs:205:9
```

No OOMs detected on the job.

**Comment by philip-stoev** [comment]
Also seen as #16793  . IT happens twice a day on average in the CI, so I am setting this as a release blocker.

**Comment by uce** [comment]
As discussed on [Slack](https://materializeinc.slack.com/archives/CTESPM7FU/p1677483595783769), we're moving this from release blocker to p0.

**Comment by philip-stoev** [comment]
Happened again in Limits https://buildkite.com/materialize/nightlies/builds/1933#0186c8f1-9b16-4cba-aa7a-0c73c3fa00fe

**Comment by philip-stoev** [comment]
And again https://buildkite.com/materialize/nightlies/builds/1987#0186ecc3-e79b-44e2-99b9-b7cdf81e533a

**Comment by aljoscha** [comment]
@pH14 hrm, I don't think I know what's what anymore. 🤯

**Comment by danhhz** [comment]
Hey @pH14 I went to close this out, but seems like sentry has some from as late as version v0.51.0. I see #17877 went into v0.47.0, were there any more fixes after that?

**Comment by pH14** [comment]
None that I can think of. Been looking at this for a short while, but cannot tell you why we still saw this on 0.51.0, nor why we haven't seen it since

**Comment by danhhz** [comment]
cool! marked this as resolved as of v0.54.0 in sentry, which I think means it will reopen the github issue if we see it again, so closing this as fixed

**Comment by def-** [comment]
Seen again today in Sentry: https://materializeinc.sentry.io/issues/3748042351/

**Comment by def-** [comment]
Also happened in parallel-workload: https://buildkite.com/materialize/nightlies/builds/3831#018a6a4c-ce81-4d31-8197-a16f8efe979f
@pH14 Any plans on working on this? Maybe we can provide more info in the logs to make it easier to understand this as a next step.

**Comment by bkirwi** [comment]
@def- - currently believe the recent parallel-workload failures are a distinct issue that is fixed by https://github.com/MaterializeInc/materialize/pull/21655... so while it's its own thing, hopefully it won't live long enough to deserve a new issue!

**Comment by bkirwi** [comment]
The fix has merged and I don't see the same failure rate in recent runs, so I'm declaring the recent regression fixed. No opinion on the sentry issue from June.

**Comment by def-** [comment]
@pH14 This just panicked in production again ^

**Comment by danhhz** [comment]
This was on the same incident-70 pod that's having all the read lease timeout issues, and this is exactly what the read leases are these for, so I think we're fine to ignore this particular occurrence.

**Comment by benesch** [comment]
We haven't seen this in a while, as far as I'm aware. Going to optimistically close this. persist is tracking the "leaseless readers" in its mega tracking issue.
